api/ipfsproxy/ipfsproxy.go

```diff
@@ -0,0 +1,467 @@
+package ipfsproxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	rpc "github.com/hsanjuan/go-libp2p-gorpc"
+	cid "github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	peer "github.com/libp2p/go-libp2p-peer"
+	madns "github.com/multiformats/go-multiaddr-dns"
+	manet "github.com/multiformats/go-multiaddr-net"
+
+	"github.com/ipfs/ipfs-cluster/adder/adderutils"
+	"github.com/ipfs/ipfs-cluster/api"
+	"github.com/ipfs/ipfs-cluster/rpcutil"
+)
+
+// DNSTimeout is used when resolving DNS multiaddresses in this module
+var DNSTimeout = 5 * time.Second
+
+var logger = logging.Logger("ipfsproxy")
+
+// IPFSProxy offers an IPFS API, hijacking some interesting requests
+// and forwarding the rest to the ipfs daemon
+// it proxies HTTP requests to the configured IPFS
+// daemon. It is able to intercept these requests though, and
+// perform extra operations on them.
+type IPFSProxy struct {
+	ctx    context.Context
+	cancel func()
+
+	config   *Config
+	nodeAddr string
+
+	rpcClient *rpc.Client
+	rpcReady  chan struct{}
+
+	listener net.Listener // proxy listener
+	server   *http.Server // proxy server
+
+	shutdownLock sync.Mutex
+	shutdown     bool
+	wg           sync.WaitGroup
+}
+
+type ipfsError struct {
+	Message string
+}
+
+type ipfsPinType struct {
+	Type string
+}
+
+type ipfsPinLsResp struct {
+	Keys map[string]ipfsPinType
+}
+
+type ipfsPinOpResp struct {
+	Pins []string
+}
+
+// From https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L49
+type ipfsAddResp struct {
+	Name  string
+	Hash  string `json:",omitempty"`
+	Bytes int64  `json:",omitempty"`
+	Size  string `json:",omitempty"`
+}
+
+// NewIPFSProxy returns and IPFSProxy component
+func NewIPFSProxy(cfg *Config) (*IPFSProxy, error) {
```

Comment:
We can rename this to `New(..) (*Server, error)`.

Then the methods for the server object can be like `func(proxy *Server) Name(...)`.